# Contents
### 1. Importing Libraries
### 2. Importing Data
### 3. Drop 'eval_set Column'
### 4. Rename 'order_dow' to 'orders_day_of_week'
### 5. Import 'Departments.csv'
### 6. Transpose Departments
### 7. Fix Header
### 8. Create Data Dictionary for Dept. # and Product Type
### 9. Create df_snacks from products dept. # using data dict
### 10. Change column heading 'order_number' to '#_orders_from_user'
### 11. Which hour of the day has the most orders
### 12. Create df of breakfast foods
### 13. Create df of departments that contain dinner party foods

# 1. Importing Libraries

In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 2. Importing Data

In [2]:
# Creating path to raw data folder
path = r'C:\Documents\Career Foundry\Python\Instacart Basket Analysis\02 Data\Original Data'

In [3]:
path

'C:\\Documents\\Career Foundry\\Python\\Instacart Basket Analysis\\02 Data\\Original Data'

In [5]:
# Importing Dataframes
df_orders = pd.read_csv(os.path.join(path, 'orders.csv'), index_col = False)

In [6]:
df_orders.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [7]:
df_prod = pd.read_csv(os.path.join(path, 'products.csv'), index_col = False)

In [8]:
df_prod.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# Exercise Work

# 3. Drop 'eval_set Column'

In [80]:
# accidentally ran this twice while trying to add a comment
# deleted column 'eval_set' from df_orders
df_orders.drop(columns = ['eval_set'])

KeyError: "['eval_set'] not found in axis"

In [10]:
df_orders['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

# 4. Rename 'order_dow' to 'orders_day_of_week'

In [11]:
# Rename 'order_dow' to 'orders_day_of_week'
df_orders.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [12]:
df_orders.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [13]:
df_orders['order_id'] = df_orders['order_id'].astype('str')

In [14]:
df_orders['order_id'].dtype

dtype('O')

# 5. Import 'Departments.csv'

In [ ]:
df_dep = pd.read_csv(os.path.join(path, 'departments.csv'), index_col = False)

In [16]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

# 6. Transpose Departments

In [17]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [18]:
df_dep_t = df_dep.T

In [19]:
df_dep_t.head()

0
department_id  department
1                  frozen
2                   other
3                  bakery
4                 produce

In [20]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

# 7. Fix Header

In [21]:
new_header = df_dep_t.iloc[0]

In [22]:
new_header


0    department
Name: department_id, dtype: object

In [23]:
df_dep_t_new = df_dep_t[1:]

In [24]:
df_dep_t_new


0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [25]:
df_dep_t_new.columns = new_header

In [26]:
df_dep_t_new


department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [27]:
df_prod.head(20)

product_id                                       product_name  aisle_id  \
0            1                         Chocolate Sandwich Cookies        61   
1            2                                   All-Seasons Salt       104   
2            3               Robust Golden Unsweetened Oolong Tea        94   
3            4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4            5                          Green Chile Anytime Sauce         5   
5            6                                       Dry Nose Oil        11   
6            7                     Pure Coconut Water With Orange        98   
7            8                  Cut Russet Potatoes Steam N' Mash       116   
8            9                  Light Strawberry Blueberry Yogurt       120   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
11          12                         Chocolate Fudge Layer Cake       119   
12          13                                  Saline Nasal Mist        11   
13          14                     Fresh Scent Dishwasher Cleaner        74   
14          15                           Overnight Diapers Size 6        56   
15          16                      Mint Chocolate Flavored Syrup       103   
16          17                                  Rendered Duck Fat        35   
17          18                Pizza for One Suprema  Frozen Pizza        79   
18          19   Gluten Free Quinoa Three Cheese & Mushroom Blend        63   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   

    department_id  prices  
0              19     5.8  
1              13     9.3  
2               7     4.5  
3               1    10.5  
4              13     4.3  
5              11     2.6  
6               7     4.4  
7               1     1.1  
8              16     7.0  
9               7     8.4  
10              7     2.8  
11              1     9.4  
12             11     1.1  
13             17     6.5  
14             18    11.2  
15             19     5.2  
16             12    17.1  
17              1    10.6  
18              9    12.0  
19              7     6.0

# 8. Create Data Dictionary for Dept. # and Product Type

In [30]:
data_dict = df_dep_t_new.to_dict('index')

In [31]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [33]:
df_prod.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [34]:
print(data_dict.get('19'))

{'department': 'snacks'}


# 9. Create df_snacks from products dept. # using data dict

In [35]:
df_snacks  = df_prod[df_prod['department_id']==19]

In [36]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [37]:
df_snacks_2 = df_prod.loc[df_prod['department_id'] == 19]

In [39]:
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [41]:
df_snacks_3 = df_prod.loc[df_prod['department_id'].isin([19])]

In [42]:
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [45]:
df_snacks_4 = df_prod.loc[df_prod['department_id'].isin([19, 17, 18])]

In [46]:
df_snacks_4.head()

product_id                             product_name  aisle_id  \
0            1               Chocolate Sandwich Cookies        61   
13          14           Fresh Scent Dishwasher Cleaner        74   
14          15                 Overnight Diapers Size 6        56   
15          16            Mint Chocolate Flavored Syrup       103   
24          25  Salted Caramel Lean Protein & Fiber Bar         3   

    department_id  prices  
0              19     5.8  
13             17     6.5  
14             18    11.2  
15             19     5.2  
24             19     1.9

In [48]:
# Rename path to be able to access '02 Data\Prepped Data'
path = r'C:\Documents\Career Foundry\Python\Instacart Basket Analysis\02 Data'

In [49]:
df_orders.to_csv(os.path.join(path, 'Prepped Data', 'orders_wrangled.csv'))

In [50]:
df_orders = df_orders.drop(columns = ['eval_set'])

In [51]:
df_orders.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [52]:
df_orders.to_csv(os.path.join(path, 'Prepped Data', 'orders_wrangled.csv'))

# Task Work

In [57]:
df_orders.dtypes

order_id                   object
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [58]:
# change type of 'user_id' to object because it doesn't need to be manipulated as a number
df_orders['user_id'] = df_orders['user_id'].astype ('str')

In [59]:
df_orders['user_id'].dtype

dtype('O')

# 10. Change column heading 'order_number' to '#_orders_from_user'

In [60]:
df_orders.rename(columns = {'order_number' : '#_orders_from_user'}, inplace = False)

order_id user_id  #_orders_from_user  orders_day_of_week  \
0        2539329       1                   1                   2   
1        2398795       1                   2                   3   
2         473747       1                   3                   3   
3        2254736       1                   4                   4   
4         431534       1                   5                   4   
...          ...     ...                 ...                 ...   
3421078  2266710  206209                  10                   5   
3421079  1854736  206209                  11                   4   
3421080   626363  206209                  12                   1   
3421081  2977660  206209                  13                   1   
3421082   272231  206209                  14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [61]:
df_orders.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 11. Which hour of the day has the most orders

In [62]:
# count the frequency of 'order_hour_of_day'
df_orders['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

## The busiest time for placing orders is from 10-11 AM

In [63]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

## The value of 4 in the 'department_id' column in the df_prods dataframe is produce

# 12. Create df of breakfast foods

In [65]:
# Breakfast items are '14' in 'department_id' in df_prod
df_breakfast = df_prod[df_prod['department_id'] == 14]

In [66]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

# 13. Create df of departments that contain dinner party foods

In [70]:
# Filter dinner party items using 'department_id' in df_prod
# '5': {'department': 'alcohol'}, '20': {'department': 'deli'}, '7': {'department': 'beverages'}, '12': {'department': 'meat seafood'}
df_dinner_party = df_prod.loc[df_prod['department_id'].isin([5,20,7,12])]

In [71]:
df_dinner_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

## The last dataframe I created, df_dinner_party, has 7650 rows

In [75]:
# Information about 'user_id' '1'
df_user_1 = df_orders.loc[df_orders['user_id'] == '1']

In [76]:
df_user_1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

## User 1 has made 11 orders.  3 orders were made on the 1st day of the week, 2 on the 2nd, 2 on the 3rd, 4 on the 4th and 0 on the 5th, 6th & 7th. 7 orders were made in the morning (between 7 & 10 AM) and 4 were made in the afternoon (between 12 & 5 pm). 

In [81]:
# Export df_orders as "orders_wrangled.csv"
df_orders.to_csv(os.path.join(path, 'Prepped Data', 'orders_wrangled.csv'))

In [82]:
# Export df_dep_t_new as "departments_wrangled.csv"
df_dep_t_new.to_csv(os.path.join(path, 'Prepped Data', 'departments_wrangled.csv'))